share/getter.go

```diff
@@ -0,0 +1,98 @@
+package share
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/minio/sha256-simd"
+
+	"github.com/celestiaorg/celestia-node/share/ipld"
+
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+// Getter interface provides a set of accessors for shares by the Root.
+// Automatically verifies integrity of shares(exceptions possible depending on the implementation)
+type Getter interface {
+	// GetShare gets a Share by coordinates in EDS.
+	GetShare(ctx context.Context, root *Root, row, col int) (Share, error)
+
+	// GetEDS gets the full EDS identified by the given root.
+	GetEDS(context.Context, *Root) (*rsmt2d.ExtendedDataSquare, error)
+
+	// GetSharesByNamespace gets all shares from an EDS within the given namespace.
+	// Shares are returned in a row-by-row order if the namespace spans multiple rows.
+	GetSharesByNamespace(context.Context, *Root, namespace.ID) (NamespaceShares, error)
+}
+
+// NamespaceShares represents all the shares with proofs of a specific namespace of an EDS.
+type NamespaceShares []RowNamespaceShares
+
+// RowNamespaceShares represents all the shares with proofs of a specific namespace within a Row of
+// an EDS.
+type RowNamespaceShares struct {
+	Shares []Share
+	Proof  *ipld.Proof
+}
+
+// Flatten returns the concatenated slice of all RowNamespaceShares shares.
+func (ns NamespaceShares) Flatten() []Share {
+	shares := make([]Share, 0)
+	for _, row := range ns {
+		shares = append(shares, row.Shares...)
+	}
+	return shares
+}
+
+// Verify validates NamespaceShares by checking every row with nmt inclusion proof.
+func (ns NamespaceShares) Verify(root *Root, nID namespace.ID) error {
```

Comment:
why did this have to be part of this PR cc @walldiss @distractedm1nd 

can we extract into a follow-up?

Reply:
Its just part of the new type, and is tested in this PR - we could extract it but it would add another PR to the stack needed for shrex/nd.